### ISSUE TITLE
`assigning_clones` false negative with `str::to_string`

### ISSUE THREAD

**Issue by xtqqczze** [rust-lang/rust-clippy#16517]
### Summary

The [`assigning_clones`](https://rust-lang.github.io/rust-clippy/master/index.html#assigning_clones) lint fails to trigger when assigning the result of `str::to_string()`.

### Lint Name

`assigning_clones`

### Reproducer

I tried this code:

```rust
pub fn m1(a: &str) -> String {
    let mut b = String::new();
    b = a.to_string();
    b
}
```

https://godbolt.org/z/snT9h8ea6

I expected to see this happen:

```
warning: assigning the result of `str::to_string()` may be inefficient
```

Instead, this happened:

Lint did not trigger

### Version

```text
nightly
```

**Comment by profetia** [comment]
This should be fixed easily reusing the logic in #16329, after it is merged.

**Comment by estebank** [comment]
CC https://github.com/rust-lang/rust-clippy/issues/16517
